include/IECoreScene/PrimitiveVariableAlgos.h

```diff
@@ -0,0 +1,299 @@
+template<typename U, typename I = int>
+class DeleteFlaggedUniformFunctor : public DeleteFlagged<U, I>
+{
+	public:
+
+		DeleteFlaggedUniformFunctor( const PrimitiveVariable::IndexedView<U> &deleteFlag, bool invert ) : DeleteFlagged<U, I>( deleteFlag, invert )
+		{
+		}
+
+		template<typename T, template<typename> class V>
+		IndexedData operator()( const V<std::vector<T> > *data )
+		{
+			const std::vector<T> &inputs = data->readable();
+			IECoreScene::PrimitiveVariable::IndexedView<T> dataView( inputs, this->m_dataIndices );
+
+			IndexedPrimitiveVariableBuilder<T, V> builder( inputs.size(), this->m_dataIndices ? this->m_dataIndices->size() : 0 );
+
+			for( size_t i = 0; i < dataView.size(); ++i )
+			{
+				if( this->shouldKeepPrimitive( i ) )
+				{
+					builder.addIndexedValue( dataView, i );
+				}
+			}
+
+			return builder.indexedData();
+		}
+
+		IndexedData operator()( const IECore::Data *data )
+		{
+			throw IECore::Exception(
+				boost::str( boost::format( "Unexpected Data: %1%" ) % ( data ? data->typeName() : std::string( "nullptr" ) ) )
+			);
+		}
+
+};
+
+template<typename U, typename I = int>
+class DeleteFlaggedVertexFunctor : public DeleteFlagged<U, I>
+{
+	public:
+
+		DeleteFlaggedVertexFunctor( const PrimitiveVariable::IndexedView<U> &deleteFlag, IECore::ConstIntVectorDataPtr verticesPerCurve, bool invert )
+			: DeleteFlagged<U, I>( deleteFlag, invert ), m_verticesPerCurve( verticesPerCurve )
+		{
+		}
+
+		template<typename T, template<typename> class V>
+		IndexedData operator()( const V<std::vector<T> > *data )
+		{
+			const std::vector<T> &inputs = data->readable();
+			const std::vector<int> &verticesPerCurve = m_verticesPerCurve->readable();
+
+			IECoreScene::PrimitiveVariable::IndexedView<T> dataView( inputs, this->m_dataIndices );
+			IndexedPrimitiveVariableBuilder<T, V> builder( inputs.size(), this->m_dataIndices ? this->m_dataIndices->size() : 0 );
+
+			size_t offset = 0;
+			for( size_t c = 0; c < verticesPerCurve.size(); ++c )
+			{
+				int numVerts = verticesPerCurve[c];
+				if( this->shouldKeepPrimitive( c ) )
+				{
+					for( int v = 0; v < numVerts; ++v )
+					{
+						builder.addIndexedValue( dataView, offset + v );
+					}
+				}
+				offset += numVerts;
+			}
+
+			return builder.indexedData();
+		}
+
+		IndexedData operator()( const IECore::Data *data )
+		{
+			throw IECore::Exception(
+				boost::str( boost::format( "Unexpected Data: %1%" ) % ( data ? data->typeName() : std::string( "nullptr" ) ) )
+			);
+		}
+
+	private:
+
+		IECore::ConstIntVectorDataPtr m_verticesPerCurve;
+
+};
+
+template<typename U, typename I = int>
+class DeleteFlaggedVaryingFunctor : public DeleteFlagged<U, I>
+{
+	public:
+
+		DeleteFlaggedVaryingFunctor( const PrimitiveVariable::IndexedView<U> &deleteFlag, const CurvesPrimitive *curvesPrimitive, bool invert )
+			: DeleteFlagged<U, I>( deleteFlag, invert ), m_curvesPrimitive( curvesPrimitive )
+		{
+		}
+
+		template<typename T, template<typename> class V>
```

Comment:
What's the purpose of the `V` template parameter? My guess would be it's intended to distinguish between `GeometricTypedData` and `TypedData`, but I can't see anywhere it's having an effect. Is there an opportunity to simplify here?

Reply:
It's passed to the template class which builds the output primvar

Reply:
I tried adding two functions per functor (without the enable_if), one for TypedData and another for GeometricTypedData and another function for the base functionality, but the template template parameter seemed the better approach here.